offer-manager-app/src/main/groovy/life/qbic/portal/offermanager/dataresources/projects/ProjectMainConnector.groovy

```diff
@@ -0,0 +1,136 @@
+
+import java.sql.Connection
+import java.sql.PreparedStatement
+import java.sql.ResultSet
+import java.sql.Statement
+
+import life.qbic.openbis.openbisclient.OpenBisClient
+
+import ch.ethz.sis.openbis.generic.asapi.v3.dto.common.operation.IOperation;
+import ch.ethz.sis.openbis.generic.asapi.v3.dto.project.create.CreateProjectsOperation;
+import ch.ethz.sis.openbis.generic.asapi.v3.dto.project.create.ProjectCreation;
+import ch.ethz.sis.openbis.generic.asapi.v3.dto.project.id.ProjectIdentifier;
+import ch.ethz.sis.openbis.generic.asapi.v3.dto.space.create.CreateSpacesOperation;
+import ch.ethz.sis.openbis.generic.asapi.v3.dto.space.create.SpaceCreation;
+import ch.ethz.sis.openbis.generic.asapi.v3.dto.space.id.SpacePermId;
+import ch.ethz.sis.openbis.generic.asapi.v3.IApplicationServerApi;
+
+/**
+ * Provides operations on QBiC project data
+ *
+ * This class implements the data sources of the different use cases and is responsible for transferring data to the customer db and openBIS
+ *
+ * @since: 1.0.0
+ *
+ */
+@Log4j2
+class ProjectMainConnector implements CreateProjectDataSource, CreateProjectSpaceDataSource {
+
+  /**
+   * A connection to the customer database used to create queries.
+   */
+  private final ProjectDbConnector projectDbConnector
+  private final OpenBisClient openbisClient
+
+  /**
+   * Constructor for a ProjectOpenBisAndDBConnector
+   * @param connection a connection to the customer db
+   * 
+   * @see Connection
+   */
+  ProjectMainConnector(ProjectDbConnector projectDbConnector, OpenBisClient openbisClient) {
+    this.projectDbConnector = projectDbConnector
+    this.openbisClient = openbisClient
+  }
+
+  @Override
+  void createProjectSpace(ProjectSpace projectSpace) throws ProjectSpaceExistsException, DatabaseQueryException {
+    String spaceName = projectSpace.getName()
+    if(openbisClient.spaceExists(spaceName)) {
+      throw new ProjectSpaceExistsException("Project space "+spaceName+" could not be created, as it exists in openBIS already!")
+    }
+    try {
+      //we don't provide a description in our data model for now, but it's optional anyway
+      createOpenbisSpace(spaceName, "")
+
+    } catch (Exception e) {
+      log.error(e.message)
+      log.error(e.stackTrace.join("\n"))
+      throw new DatabaseQueryException("Could not create project space.")
+    }
+  }
+
+  private void createOpenbisSpace(String spaceName, String description) {
+    // TODO this is not how the api should be used, I reckon
+    IApplicationServerApi api = openbisClient.getV3()
+    SpaceCreation space = new SpaceCreation()
+    space.setCode(spaceName)
+
+    space.setDescription(description);
+
+    IOperation operation = new CreateSpacesOperation(space)
+    api.handleOperations(operation)
+  }
+
+  private void createOpenbisProject(ProjectSpace space, ProjectCode projectCode, String description) {
+    // TODO this is not how the api should be used, I reckon
+    IApplicationServerApi api = openbisClient.getV3()
```

Comment:
Can you elaborate?

Reply:
I think we discussed this yesterday and agreed to use it like this for now, I forgot to remove the TODO

Reply:
yes, i was wondering about the comment